core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -637,11 +639,12 @@ private[log] class Cleaner(val id: Int,
    * @param segment The segment to index
    * @param map The map in which to store the key=>offset mapping
    *
-   * @return The final offset covered by the map
+   * @return The final offset covered by the map or -1 if the map is full
    */
   private def buildOffsetMapForSegment(topicAndPartition: TopicAndPartition, segment: LogSegment, map: OffsetMap): Long = {
     var position = 0
     var offset = segment.baseOffset
+    val maxDesiredMapSize = (map.slots * this.dupBufferLoadFactor).toInt
```

Comment:
We found that if we fill the map to the very last byte, then the method SkimpyOffsetMap.get(...) may loop over so many attempts to get into integer overflow ! and that was slow. 
so retaining a load factor seemed good. 
We may open a jira bug for that overflow though.